dace/frontend/python/newast.py

```diff
@@ -1970,46 +1971,116 @@ def visit_For(self, node: ast.For):
                 laststate, first_loop_state, end_loop_state, indices[0],
                 ranges[0][0],
                 '%s %s %s' % (indices[0], loop_cond, ranges[0][1]),
                 '%s + %s' % (indices[0], ranges[0][2]), last_loop_state)
+            incr = {indices[0]: '%s + %s' % (indices[0], ranges[0][2])}
+            continue_states = self.continue_states.pop()
+            while continue_states:
+                next_state = continue_states.pop()
+                out_edges = self.sdfg.out_edges(next_state)
+                for e in out_edges:
+                    self.sdfg.remove_edge(e)
+                self.sdfg.add_edge(next_state, loop_guard,
+                                   dace.InterstateEdge(assignments=incr))
+            break_states = self.break_states.pop()
+            while break_states:
+                next_state = break_states.pop()
+                out_edges = self.sdfg.out_edges(next_state)
+                for e in out_edges:
+                    self.sdfg.remove_edge(e)
+                self.sdfg.add_edge(next_state, loop_end, dace.InterstateEdge())
+            self.loop_idx -= 1
         else:
             raise DaceSyntaxError(
                 self, node, 'Unsupported for-loop iterator "%s"' % iterator)
 
     def visit_While(self, node: ast.While):
         # Add an initial loop state with a None last_state (so as to not
         # create an interstate edge)
+        self.loop_idx += 1
+        self.continue_states.append([])
+        self.break_states.append([])
         laststate, first_loop_state, last_loop_state = \
             self._recursive_visit(node.body, 'while', node.lineno)
         end_loop_state = self.last_state
 
         # Add loop to SDFG
         loop_cond = astutils.unparse(node.test)
-        self.sdfg.add_loop(laststate, first_loop_state, end_loop_state, None,
-                           None, loop_cond, None, last_loop_state)
+        _, loop_guard, loop_end = self.sdfg.add_loop(
+            laststate, first_loop_state, end_loop_state, None,
+            None, loop_cond, None, last_loop_state)
+        
+        continue_states = self.continue_states.pop()
+        while continue_states:
+            next_state = continue_states.pop()
+            out_edges = self.sdfg.out_edges(next_state)
+            for e in out_edges:
+                self.sdfg.remove_edge(e)
+            self.sdfg.add_edge(next_state, loop_guard, dace.InterstateEdge())
+        break_states = self.break_states.pop()
+        while break_states:
+            next_state = break_states.pop()
+            out_edges = self.sdfg.out_edges(next_state)
+            for e in out_edges:
+                self.sdfg.remove_edge(e)
+            self.sdfg.add_edge(next_state, loop_end, dace.InterstateEdge())
+        self.loop_idx -= 1
+
+    def visit_Break(self, node: ast.Break):
+        if self.loop_idx < 0:
+            error_msg = "'break' is only supported inside for and while loops "
+            if self.nested:
```

Comment:
Yes, this works (just tried). It doesn't work (and hopefully returns an informative error msg) when `break` or `continue` is in a nested scope relative to their corresponding loop guard. The reason that such a thing is not supported has to do with how `break`/`continue` are currently implemented; with interstate edges. Such an approach doesn't work with nested scopes, since you cannot connect a state of the nested scope with a node/state of the outer scope. This is also why Maps are not supported (apart from the semantical issue).